### PR TITLE
add scripts and templates to deploy in aws

### PR DESCRIPTION
**Pull request by champagnearden**
Comme demandé lors de notre échange de mails, voici la version fonctionnant avec un déploiement dans aws.
Voir le dossier `aws` pour plus d'informations

### CODE REVIEW

**Comment by champagnearden** [comment]
> Hello, thank you very much for the value you're bringing to this repository !
> 
> I will review the files.
> 
> For now, just a side note, I see most of the kubernetes manifests you're using are just a duplication of the existing ones. The maintenance of the kubernetes definition will quickly become complex if duplicated. Could you please remove the already existing kubernetes manifests you're using from the aws folder and use the ones from repo-root/k8s ? It will ease the review and reduce future maintenance.
> 
> Thank you again ! Romain

Hello,
I updated to revise the code.
Now, the files that changed have been moved to `directory` and all the scripts are directly referencing the `k8s` folder.

Thanks for your remark and feel free to ping me if something went wrong !

**Review by Rbillon59 — CHANGES_REQUESTED** [review]
I did not finish the review yet, because I see you're still pushing code. Could you put the PR in draft mode until you finish to push your changes ? So I can review when it's done ?

Thank you

File: aws/README.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+- `-r` : ***eu-west-1***
+- `-p` : ***default***
+
+So, if you want to test the deployment you can run 
+```bash 
+# Install the stack rennes with 2 nodes
+./install-rennes-from-scratch.sh
+# Delete the stack rennes
+./remove-rennes.sh
+```
```
[the remark this answers is not in the document]
Comment (champagnearden) [reply aws/README.md:153]: Ok I removed them !
Therefore, the scripts are calling other ones in the `scripts` folder, so you have to make sure that the scripts are run from the script folder.
bash take the path given in scripts for the user:
If you're in the root folder and run `./generateCloudForm.sh`, the following lines will throw an error
```bash
cp ../aws-files/grafana/* ../../k8s/tool/grafana
cp ../aws-files/jmeter/* ../../k8s/jmeter
sed -i "s/##VolumeId##/${volumes[0]}/g" ../../k8s/jmeter/jmeter-pv.yaml
cd ../../
kubectl create -R -f k8s/
```

File: aws/aws-files/grafana/grafana-deployment.yaml
[the remark this answers is not in the document]
Comment (champagnearden) [reply aws/aws-files/grafana/grafana-deployment.yaml:1]: from lines 33 to 40, I had to change the paths from the original version

File: aws/scripts/run-scenarii-jmeter-tool-to-AWS.sh
Comment (Rbillon59) [inline aws/scripts/run-scenarii-jmeter-tool-to-AWS.sh:1]: Why is this file needed ?  I don't really see something aws related here.

Also, the `./start_test.sh -j ${scenario} -n ${namespace} -c -m -i ${injectors} -r` may fail if the user is not using modules and csv. So at the end, you would need the exact same parameters as the `start_test.sh`.

File: start_test.sh [hunk 2]
```diff
@@ -108,15 +108,15 @@ logit "INFO" "Waiting for master pod to be available"
...
+logit "INFO" "master_pod is ${master_pod}"
 
 #Get Slave pod details
 slave_pods=($(kubectl get pods -n "${namespace}" | grep jmeter-slave | grep Running | awk '{print $1}'))
 slave_num=${#slave_pods[@]}
 slave_digit="${#slave_num}"
 
 # jmeter directory in pods
-jmeter_directory="/opt/jmeter/apache-jmeter/bin"
+jmeter_directory="//opt/jmeter/apache-jmeter/bin"
```
Comment (Rbillon59) [inline start_test.sh:119]:
```suggestion
jmeter_directory="/opt/jmeter/apache-jmeter/bin"
```

File: start_test.sh [hunk 3]
```diff
@@ -147,12 +147,12 @@ logit "INFO" "Number of slaves is ${slave_num}"
 for ((i=0; i<end; i++))
 do
     logit "INFO" "Copying scenario/${jmx_dir}/${jmx} to ${slave_pods[$i]}"
-    kubectl cp -c jmslave "scenario/${jmx_dir}/${jmx}" -n "${namespace}" "${slave_pods[$i]}:/opt/jmeter/apache-jmeter/bin/" &
+    kubectl cp -c jmslave "scenario/${jmx_dir}/${jmx}" -n "${namespace}" "${slave_pods[$i]}://opt/jmeter/apache-jmeter/bin/" &
```
Comment (Rbillon59) [inline start_test.sh:150]:
```suggestion
    kubectl cp -c jmslave "scenario/${jmx_dir}/${jmx}" -n "${namespace}" "${slave_pods[$i]}:/opt/jmeter/apache-jmeter/bin/" &
```

File: start_test.sh [hunk 4]
```diff
@@ -147,12 +147,12 @@ logit "INFO" "Number of slaves is ${slave_num}"
...
 done # for i in "${slave_pods[@]}"
 logit "INFO" "Finish copying scenario in slaves pod"
 
 logit "INFO" "Copying scenario/${jmx_dir}/${jmx} into ${master_pod}"
-kubectl cp -c jmmaster "scenario/${jmx_dir}/${jmx}" -n "${namespace}" "${master_pod}:/opt/jmeter/apache-jmeter/bin/" &
+kubectl cp -c jmmaster "scenario/${jmx_dir}/${jmx}" -n "${namespace}" "${master_pod}://opt/jmeter/apache-jmeter/bin/" &
```
Comment (Rbillon59) [inline start_test.sh:155]:
```suggestion
kubectl cp -c jmmaster "scenario/${jmx_dir}/${jmx}" -n "${namespace}" "${master_pod}:/opt/jmeter/apache-jmeter/bin/" &
```

File: start_test.sh [hunk 5]
```diff
@@ -208,8 +208,8 @@ wait
 for ((i=0; i<end; i++))
 do
         logit "INFO" "Starting jmeter server on ${slave_pods[$i]} in parallel"
-        kubectl cp -c jmslave "scenario/${jmx_dir}/jmeter_injector_start.sh" -n "${namespace}" "${slave_pods[$i]}:/opt/jmeter/jmeter_injector_start"
-        kubectl exec -c jmslave -i -n "${namespace}" "${slave_pods[$i]}" -- /bin/bash "/opt/jmeter/jmeter_injector_start" &  
+        kubectl cp -c jmslave "scenario/${jmx_dir}/jmeter_injector_start.sh" -n "${namespace}" "${slave_pods[$i]}://opt/jmeter/jmeter_injector_start"
```
Comment (Rbillon59) [inline start_test.sh:211]:
```suggestion
        kubectl cp -c jmslave "scenario/${jmx_dir}/jmeter_injector_start.sh" -n "${namespace}" "${slave_pods[$i]}:/opt/jmeter/jmeter_injector_start"
```

File: start_test.sh [hunk 6]
```diff
@@ -208,8 +208,8 @@ wait
...
+        kubectl exec -c jmslave -i -n "${namespace}" "${slave_pods[$i]}" -- //bin/bash "//opt/jmeter/jmeter_injector_start" &
```
Comment (Rbillon59) [inline start_test.sh:212]:
```suggestion
        kubectl exec -c jmslave -i -n "${namespace}" "${slave_pods[$i]}" -- //bin/bash "/opt/jmeter/jmeter_injector_start" &  
```

File: start_test.sh [hunk 7]
```diff
@@ -236,25 +236,20 @@ echo "slave_array=(${slave_array[@]}); index=${slave_num} && while [ \${index} -
...
     echo "jmeter ${param_host} ${param_user} ${report_command_line} --logfile /report/${jmx}_$(date +"%F_%H%M%S").jtl --nongui --testfile ${jmx} -Dserver.rmi.ssl.disable=true --remoteexit --remotestart ${slave_list} >> jmeter-master.out 2>> jmeter-master.err &"
     echo "trap 'kill -10 1' EXIT INT TERM"
     echo "java -jar /opt/jmeter/apache-jmeter/lib/jolokia-java-agent.jar start JMeter >> jmeter-master.out 2>> jmeter-master.err"
-    echo "echo \"Starting load test at : $(date)\" && wait"
+    echo "wait"
 } >> "scenario/${jmx_dir}/load_test.sh"
 
-logit "INFO" "Copying scenario/${jmx_dir}/load_test.sh into  ${master_pod}:/opt/jmeter/load_test"
-kubectl cp -c jmmaster "scenario/${jmx_dir}/load_test.sh" -n "${namespace}" "${master_pod}:/opt/jmeter/load_test"
+logit "INFO" "Copying scenario/${jmx_dir}/load_test.sh into  ${master_pod}://opt/jmeter/load_test"
```
Comment (Rbillon59) [inline start_test.sh:245]:
```suggestion
logit "INFO" "Copying scenario/${jmx_dir}/load_test.sh into  ${master_pod}:/opt/jmeter/load_test"
```

File: start_test.sh [hunk 8]
```diff
@@ -236,25 +236,20 @@ echo "slave_array=(${slave_array[@]}); index=${slave_num} && while [ \${index} -
...
+kubectl cp -c jmmaster "scenario/${jmx_dir}/load_test.sh" -n "${namespace}" "${master_pod}://opt/jmeter/load_test"
```
Comment (Rbillon59) [inline start_test.sh:246]:
```suggestion
kubectl cp -c jmmaster "scenario/${jmx_dir}/load_test.sh" -n "${namespace}" "${master_pod}:/opt/jmeter/load_test"
```

File: start_test.sh [hunk 9]
```diff
@@ -236,25 +236,20 @@ echo "slave_array=(${slave_array[@]}); index=${slave_num} && while [ \${index} -
...
-logit "INFO" " PASSWORD : ${GRAFANA_PASSWORD}"
-logit "INFO" "################################################"
-
+logit "INFO" "${namespace} ${master_pod}"
+kubectl port-forward $(kubectl get pod | grep grafana |awk '{print $1}') 3000 &
+logit "INFO" "Connect on localhost:3000"
+logit "WARNING" "DO NOT SHARE THESES CREDENTIALS"
+logit "WARNING" "username: admin"
+logit "WARNING" "password: XhXUdmQ576H6e7"
+kubectl exec -c jmmaster -i -n "${namespace}" "${master_pod}" -- //bin/bash "//opt/jmeter/load_test"
```
Comment (Rbillon59) [inline start_test.sh:255]: The master pod does not need to be launched manually, the pod will automatically start the test. From the pod command : `until [ -f /opt/jmeter/load_test ]; do sleep 5; done && /bin/bash /opt/jmeter/load_test"]`

If you want the end of the script to follow the master logs, I would prefer you to add a command suggestion using something like `kubectl logs ...`

File: start_test.sh [hunk 10]
```diff
@@ -236,25 +236,20 @@ echo "slave_array=(${slave_array[@]}); index=${slave_num} && while [ \${index} -
...
-logit "INFO" "##################################################"
-logit "INFO" "Also using Grafana : kubectl ${namespace} port-forward svc/grafana 8443:443"
-GRAFANA_LOGIN=$(kubectl -n "${namespace}" get secret grafana-creds -o yaml | grep GF_SECURITY_ADMIN_USER: | awk -F" " '{print $2}' | base64 --decode)
-GRAFANA_PASSWORD=$(kubectl -n "${namespace}" get secret grafana-creds -o yaml | grep GF_SECURITY_ADMIN_PASSWORD: | awk -F" " '{print $2}' | base64 --decode)
-logit "INFO" " LOGIN : ${GRAFANA_LOGIN}"
-logit "INFO" " PASSWORD : ${GRAFANA_PASSWORD}"
-logit "INFO" "################################################"
-
+logit "INFO" "${namespace} ${master_pod}"
+kubectl port-forward $(kubectl get pod | grep grafana |awk '{print $1}') 3000 &
```
Comment (Rbillon59) [inline start_test.sh:250]: I like the port forward idea on this script thanks !

But you should also pass the namespace into that command if the grafana pod is namespaced

File: start_test.sh [hunk 11]
```diff
@@ -236,25 +236,20 @@ echo "slave_array=(${slave_array[@]}); index=${slave_num} && while [ \${index} -
...
+logit "INFO" "Connect on localhost:3000"
+logit "WARNING" "DO NOT SHARE THESES CREDENTIALS"
+logit "WARNING" "username: admin"
+logit "WARNING" "password: XhXUdmQ576H6e7"
```
Comment (Rbillon59) [inline start_test.sh]: The warning about sharing the credentials is a good addition, thanks.

But why removing the dynamically retrieval of the grafana creds ? If somebody want to change the grafana creds (like using external secrets to avoid having there in clear text in the repo), it will not be user friendly

File: start_test.sh [hunk 12]
```diff
@@ -236,25 +236,20 @@ echo "slave_array=(${slave_array[@]}); index=${slave_num} && while [ \${index} -
...
-logit "INFO" "Also using Grafana : kubectl ${namespace} port-forward svc/grafana 8443:443"
-GRAFANA_LOGIN=$(kubectl -n "${namespace}" get secret grafana-creds -o yaml | grep GF_SECURITY_ADMIN_USER: | awk -F" " '{print $2}' | base64 --decode)
-GRAFANA_PASSWORD=$(kubectl -n "${namespace}" get secret grafana-creds -o yaml | grep GF_SECURITY_ADMIN_PASSWORD: | awk -F" " '{print $2}' | base64 --decode)
-logit "INFO" " LOGIN : ${GRAFANA_LOGIN}"
-logit "INFO" " PASSWORD : ${GRAFANA_PASSWORD}"
-logit "INFO" "################################################"
-
+logit "INFO" "${namespace} ${master_pod}"
+kubectl port-forward $(kubectl get pod | grep grafana |awk '{print $1}') 3000 &
+logit "INFO" "Connect on localhost:3000"
```
Comment (Rbillon59) [inline start_test.sh:251]: The port forward output will be shown in the terminal, I don't think this line of log adds value

**Review by Rbillon59 — CHANGES_REQUESTED** [review]

File: start_test.sh (hunk 11 above)
Comment (champagnearden) [reply start_test.sh]: I didn't meant to remove the dynamic retrieval of grafana creds.
It was just a help line for users to have the default creds when deploying the stack

Comment (Rbillon59) [reply start_test.sh]: Using the dynamic retrieval, would the credentials be the default ones or the modified one, the user would get the most up to date value.

File: aws/scripts/generateCloudForm.sh [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+sed -i "s/##VolumeId##/${volumes[0]}/g" ../../k8s/jmeter/jmeter-pv.yaml
+cd ../../
+
+# Create the repo and the charts
+kubectl create -R -f k8s/
+helm repo add aws-ebs-csi-driver https://kubernetes-sigs.github.io/aws-ebs-csi-driver
+helm repo add jmeter-k8s-starterkit-helm-charts https://rbillon59.github.io/jmeter-k8s-starterkit-helm-chart/
+helm repo update
+helm install aws-ebs-csi-driver aws-ebs-csi-driver/aws-ebs-csi-driver
+helm install jmeter-k8s-starterkit-helm-charts/jmeter-k8s-starterkit --generate-name
```
[the remark this answers is not in the document]
Comment (champagnearden) [reply aws/scripts/generateCloudForm.sh:64]: It appears that if I dont use helm nothing seems to work...
Moreover, I needed the aws-ebs-csi-driver using helm so I used your repo too

[the remark this answers is not in the document]
Comment (Rbillon59) [reply aws/scripts/generateCloudForm.sh:64]: Yes, no doubts for the csi drivers, but I'am confused for the helm chart of the starterkit to be honest.

Actually, the `kubectl create -R k8s/` should be enough.

Did you spot what is the diff and the resources changes between the create and the helm chart deployment ? Everything should be working fine only with the create command :thinking:

**Review by Rbillon59 — APPROVED** [review]
Thank you very much for the sharing and the effort to address the reviews !
